Updates to batt_smbus driver to add functionality and work with TI BQ40Z50-R1/R2 fuel gauge

AlexKlimaj opened the pull request:
I have updated the batt_smbus driver to work with the TI BQ40Z50-R1/R2 fuel gauges. 

Voltage and current are working correctly and using updateBatteryStatus to update uORB. 

There are many more values I want to gather from the fuel gauge and get into a mavlink message. Would it be best to add them to battery_status.msg?

updateBatteryStatus in the battery driver is very geared towards battery measurements coming from the on board ADC's. I think it would be worth updating the battery_status.msg through the batt_smbus driver instead. 

BATT_SOURCE also needs to be set to 1 to bypass reading the battery ADC's in the sensors.cpp driver.

dagar commented on src/drivers/batt_smbus/batt_smbus.cpp:
```diff
@@ -602,46 +514,17 @@ BATT_SMBUS::cycle()
-			}
-		}
+		battery_status_s battery_status;
+		_battery.updateBatteryStatus(now, new_report.voltage_v, new_report.current_a,
```
I see why you did this, but I really think we need to find a better overall solution than having a low level battery driver subscribing to controller output like this.

AlexKlimaj replied:
I agree, the sensors.cpp battery update process also relies on it.

dagar commented on src/drivers/batt_smbus/batt_smbus.cpp:
```diff
@@ -1014,60 +773,47 @@ batt_smbus_main(int argc, char *argv[])
 		PX4_INFO("not started");
 		batt_smbus_usage();
-		return 1;
+		exit(1);
```
Please return an error instead of exiting. On Linux these modules are threads within a single progress, so exiting from a single module/driver kills all of px4.

AlexKlimaj replied:
Woops, I was working off of the older one and not your new commit today.

dagar commented:
@AlexKlimaj check the CI failures.

dagar commented:
@davids5 @ChristophTobler @bkueng it might be time to think about the system_power/battery_status architecture again.

AlexKlimaj commented:
PEC is not enabled on the TI BQ40Z50-R1/R2 by default. So I left it out. I will add it back in as a check.

AlexKlimaj commented:
I added PEC back in. Removed the call to update battery status which enabled me to remove the vehicle control mode and actuator controls dependency. 

I also added battery capacity to the battery uorb message. The total discharged is calculated based on the remaining capacity when the driver starts minus the current remaining capacity. 

> NuttShell (NSH)
nsh>
nsh>
nsh> batt_smbus start
batt_smbus on I2C bus 1 at 0x0b (bus: 100 KHz, max: 100 KHz)
INFO  [batt_smbus] battery found at 0xb
INFO  [batt_smbus] smart battery connected
nsh>
nsh>
nsh>
nsh> batt_sbus test
nsh: batt_sbus: command not found
nsh> batt_smbus test
INFO  [batt_smbus] V=15.17 C=0.25 DismAh=1 Cap:2495 Remaining:0.37
INFO  [batt_smbus] V=15.16 C=0.25 DismAh=1 Cap:2495 Remaining:0.37
INFO  [batt_smbus] V=15.16 C=0.25 DismAh=1 Cap:2495 Remaining:0.37

[QGC Battery Status](https://i.imgur.com/NjZNKoZ.png)

AlexKlimaj commented:
@LorenzMeier @dagar 

Added more functionality today. 

> nsh>
nsh> batt_smbus start
batt_smbus on I2C bus 1 at 0x0b (bus: 100 KHz, max: 100 KHz)
INFO  [batt_smbus] battery found at 0xb
INFO  [batt_smbus] smart battery connected
nsh>
nsh>
nsh>
nsh>
nsh> batt_smbus test
INFO  [batt_smbus] V=16.39 C=0.30 DismAh=1 Cap:2623 TempC:26.35 Remaining:0.92 RunTimeToEmpty:478 CycleCount:35 SerialNum:0001
INFO  [batt_smbus] V=16.39 C=0.30 DismAh=1 Cap:2623 TempC:26.35 Remaining:0.92 RunTimeToEmpty:478 CycleCount:35 SerialNum:0001
INFO  [batt_smbus] V=16.39 C=0.30 DismAh=1 Cap:2623 TempC:26.35 Remaining:0.92 RunTimeToEmpty:477 CycleCount:35 SerialNum:0001
INFO  [batt_smbus] V=16.39 C=0.30 DismAh=1 Cap:2623 TempC:26.35 Remaining:0.92 RunTimeToEmpty:477 CycleCount:35 SerialNum:0001
INFO  [batt_smbus] V=16.39 C=0.30 DismAh=1 Cap:2623 TempC:26.35 Remaining:0.92 RunTimeToEmpty:477 CycleCount:35 SerialNum:0001
INFO  [batt_smbus] V=16.39 C=0.30 DismAh=1 Cap:2623 TempC:26.35 Remaining:0.92 RunTimeToEmpty:477 CycleCount:35 SerialNum:0001
INFO  [batt_smbus] V=16.39 C=0.30 DismAh=1 Cap:2623 TempC:26.35 Remaining:0.92 RunTimeToEmpty:477 CycleCount:35 SerialNum:0001
INFO  [batt_smbus] V=16.39 C=0.30 DismAh=1 Cap:2623 TempC:26.35 Remaining:0.92 RunTimeToEmpty:477 CycleCount:35 SerialNum:0001
INFO  [batt_smbus] V=16.39 C=0.30 DismAh=1 Cap:2623 TempC:26.35 Remaining:0.92 RunTimeToEmpty:477 CycleCount:35 SerialNum:0001
INFO  [batt_smbus] V=16.39 C=0.30 DismAh=1 Cap:2623 TempC:26.35 Remaining:0.92 RunTimeToEmpty:477 CycleCount:35 SerialNum:0001
INFO  [batt_smbus] V=16.39 C=0.30 DismAh=1 Cap:2623 TempC:26.35 Remaining:0.92 RunTimeToEmpty:477 CycleCount:35 SerialNum:0001
INFO  [batt_smbus] V=16.39 C=0.30 DismAh=1 Cap:2623 TempC:26.35 Remaining:0.92 RunTimeToEmpty:477 CycleCount:35 SerialNum:0001
INFO  [batt_smbus] V=16.39 C=0.31 DismAh=1 Cap:2623 TempC:26.35 Remaining:0.92 RunTimeToEmpty:477 CycleCount:35 SerialNum:0001
INFO  [batt_smbus] V=16.39 C=0.31 DismAh=1 Cap:2623 TempC:26.35 Remaining:0.92 RunTimeToEmpty:477 CycleCount:35 SerialNum:0001
nsh>
nsh>
nsh>
nsh> batt_smbus test
INFO  [batt_smbus] V=16.26 C=0.31 DismAh=98 Cap:2623 TempC:29.55 Remaining:0.88 RunTimeToEmpty:455 CycleCount:35 SerialNum:0001
INFO  [batt_smbus] V=16.26 C=0.31 DismAh=98 Cap:2623 TempC:29.65 Remaining:0.88 RunTimeToEmpty:455 CycleCount:35 SerialNum:0001
INFO  [batt_smbus] V=16.26 C=0.31 DismAh=98 Cap:2623 TempC:29.65 Remaining:0.88 RunTimeToEmpty:453 CycleCount:35 SerialNum:0001
INFO  [batt_smbus] V=16.26 C=0.31 DismAh=98 Cap:2623 TempC:29.65 Remaining:0.88 RunTimeToEmpty:453 CycleCount:35 SerialNum:0001
INFO  [batt_smbus] V=16.26 C=0.31 DismAh=98 Cap:2623 TempC:29.55 Remaining:0.88 RunTimeToEmpty:453 CycleCount:35 SerialNum:0001
INFO  [batt_smbus] V=16.26 C=0.31 DismAh=98 Cap:2623 TempC:29.55 Remaining:0.88 RunTimeToEmpty:453 CycleCount:35 SerialNum:0001
INFO  [batt_smbus] V=16.26 C=0.31 DismAh=98 Cap:2623 TempC:29.55 Remaining:0.88 RunTimeToEmpty:453 CycleCount:35 SerialNum:0001
INFO  [batt_smbus] V=16.26 C=0.31 DismAh=98 Cap:2623 TempC:29.55 Remaining:0.88 RunTimeToEmpty:453 CycleCount:35 SerialNum:0001
INFO  [batt_smbus] V=16.26 C=0.31 DismAh=98 Cap:2623 TempC:29.55 Remaining:0.88 RunTimeToEmpty:453 CycleCount:35 SerialNum:0001
INFO  [batt_smbus] V=16.26 C=0.31 DismAh=98 Cap:2623 TempC:29.55 Remaining:0.88 RunTimeToEmpty:453 CycleCount:35 SerialNum:0001
INFO  [batt_smbus] V=16.26 C=0.31 DismAh=98 Cap:2623 TempC:29.55 Remaining:0.88 RunTimeToEmpty:453 CycleCount:35 SerialNum:0001
INFO  [batt_smbus] V=16.26 C=0.31 DismAh=98 Cap:2623 TempC:29.55 Remaining:0.88 RunTimeToEmpty:453 CycleCount:35 SerialNum:0001
INFO  [batt_smbus] V=16.26 C=0.31 DismAh=98 Cap:2623 TempC:29.55 Remaining:0.88 RunTimeToEmpty:453 CycleCount:35 SerialNum:0001
INFO  [batt_smbus] V=16.26 C=0.31 DismAh=98 Cap:2623 TempC:29.55 Remaining:0.88 RunTimeToEmpty:453 CycleCount:35 SerialNum:0001
nsh>
nsh>

AlexKlimaj commented:
Latest commit requires this mavlink update.

https://github.com/mavlink/mavlink/pull/841
